AbstractSQLRules2SQL.ts

```diff
@@ -0,0 +1,1165 @@
+	switch (type) {
+		case 'Alias':
+			checkArgs('Alias', rest, 2);
+			const field = matchFn(getAbstractSqlQuery(rest, 0), indent);
+			return `${field} AS "${rest[1]}"`;
+		default:
+			return matchFn(args, indent);
+	}
+};
+
+const AddBind = (bind: Binding): string => {
+	if (engine === 'postgres') {
+		if (bind[0] === 'Bind') {
+			const existingBindIndex = _.findIndex(fieldOrderings, existingBind =>
+				_.isEqual(bind, existingBind),
+			);
+			if (existingBindIndex !== -1) {
+				// Reuse the existing bind if there is one, adding 1 because the postgres binds start from $1
+				return '$' + (existingBindIndex + 1);
+			}
+		}
+		return '$' + fieldOrderings.push(bind);
+	} else {
+		fieldOrderings.push(bind);
+		return '?';
+	}
+};
+
+let fieldOrderings: Binding[] = [];
+let engine: Engines = Engines.postgres;
+
+let typeRules: Dictionary<MatchFn> = {
+	UnionQuery: (args, indent) => {
+		checkMinArgs('UnionQuery', args, 2);
+		return args
+			.map(arg => {
+				if (!_.isArray(arg)) {
+					throw new SyntaxError(
+						`Expected AbstractSqlQuery array but got ${typeof arg}`,
+					);
+				}
+				const [type, ...rest] = arg;
+				switch (type) {
+					case 'SelectQuery':
+					case 'UnionQuery':
+						return typeRules[type](rest, indent);
+					default:
+						throw new SyntaxError(`UnionQuery does not support ${type}`);
+				}
+			})
+			.join(indent + 'UNION' + indent);
+	},
+	SelectQuery: (args, indent) => {
+		const tables = [];
+		let select: string = '';
+		let where: string = '';
+		let groupBy: string = '';
+		let orderBy: string = '';
+		let limit: string = '';
+		let offset: string = '';
+		for (const arg of args) {
+			if (!_.isArray(arg)) {
+				throw new SyntaxError('`SelectQuery` args must all be arrays');
+			}
+			const [type, ...rest] = arg;
+			switch (type) {
+				case 'Select':
+					if (select !== '') {
+						throw new SyntaxError(
+							`'SelectQuery' can only accept one '${type}'`,
+						);
+					}
+					select = typeRules[type](rest, indent);
+					break;
+				case 'From':
+					tables.push(typeRules[type](rest, indent));
+					break;
+				case 'Where':
+					if (where !== '') {
+						throw new SyntaxError(
+							`'SelectQuery' can only accept one '${type}'`,
+						);
+					}
+					where = indent + typeRules[type](rest, indent);
+					break;
+				case 'GroupBy':
+					if (groupBy !== '') {
+						throw new SyntaxError(
+							`'SelectQuery' can only accept one '${type}'`,
+						);
+					}
+					groupBy = indent + typeRules[type](rest, indent);
+					break;
+				case 'OrderBy':
+					if (orderBy !== '') {
+						throw new SyntaxError(
+							`'SelectQuery' can only accept one '${type}'`,
+						);
+					}
+					orderBy = indent + typeRules[type](rest, indent);
+					break;
+				case 'Limit':
+					if (limit !== '') {
+						throw new SyntaxError(
+							`'SelectQuery' can only accept one '${type}'`,
+						);
+					}
+					limit = indent + typeRules[type](rest, indent);
+					break;
+				case 'Offset':
```

Comment:
Seems like Where, GroupBy, OrderBy, Limit & Offset could be using an abstract implementation